fili-core/src/test/java/com/yahoo/bard/webservice/async/workflows/TestAsynchronousWorkflowsBuilder.java

```diff
@@ -0,0 +1,101 @@
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+import rx.Observable;
+import rx.Observer;
+
+import java.time.Clock;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Function;
+
+/**
+ * An extension of DefaultAsynchronousWorkflowsBuilder that behaves exactly like the
+ * DefaultAsynchronousWorkflowsBuilder, except it allows outsiders to add an additional subscriber to each workflow.
+ * <p>
+ * This allows us to perform multi-response tests against the asynchronous workflow in a thread-safe manner, amongst
+ * other things.
+ */
+@Singleton
+public class TestAsynchronousWorkflowsBuilder extends DefaultAsynchronousWorkflowsBuilder {
+
+    static final Map<Workflow, Observer> SUBSCRIBERS = new HashMap<>();
+    /**
+     * A factory for constructing the asynchronous response building workflow.
+     *
+     * @param apiJobStore A service for storing and requesting job metadata
+     * @param preResponseStore A service for storing and requesting query results
+     * @param timestampGenerator The clock to use to generate timestamps
+     */
+    @Inject
+    public TestAsynchronousWorkflowsBuilder(
+            ApiJobStore apiJobStore,
+            PreResponseStore preResponseStore,
+            Clock timestampGenerator
+    ) {
+        super(apiJobStore, preResponseStore, timestampGenerator);
+    }
+
+    /**
+     * Adds the specified subscriber to the specified workflow.
+     *
+     * @param workflow  The workflow to add the countdown latch to
+     * @param workflowSubscriber  The subscriber that should be added to the specified workflow
+     */
+    public static void addSubscriber(Workflow workflow, Observer workflowSubscriber) {
+        SUBSCRIBERS.put(workflow, workflowSubscriber);
+    }
+
+    /**
+     * Clears the map of subscribers.
+     * This should be invoked in the cleanup of every test that adds latches.
+     */
+    public static void clearSubscribers() {
+        SUBSCRIBERS.clear();
+    }
+
+    @Override
+    public AsynchronousWorkflows buildAsynchronousWorkflows(
+            Observable<PreResponse> preResponseEmitter,
+            Observable<Either<PreResponse, JobRow>> payloadEmitter,
+            JobRow jobMetadata,
+            Function<JobRow, String> jobMetadataSerializer
+    ) {
+        AsynchronousWorkflows workflows = super.buildAsynchronousWorkflows(
+                preResponseEmitter,
+                payloadEmitter,
+                jobMetadata,
+                jobMetadataSerializer
+        );
+
+        Map<Workflow, Observable> workflowMap = new HashMap<>();
+        workflowMap.put(Workflow.SYNCHRONOUS, workflows.getSynchronousPayload());
+        workflowMap.put(Workflow.ASYNCHRONOUS, workflows.getAsynchronousPayload());
+        workflowMap.put(Workflow.PRERESPONSE_READY, workflows.getPreResponseReadyNotifications());
+        workflowMap.put(Workflow.JOB_MARKED_COMPLETE, workflows.getJobMarkedCompleteNotifications());
```

Comment:
Yeah, I wasn't a particularly big fan of this, either. In Java, the best approach to handle this sort of thing may be to make the `AsynchronousWorkflows` object less rigid. Give it a map that maps interfaces to `Observables`, with an enum as default implementation of the interface that enumerates the current workflows, rather than having explicit fields and getters. Then, this class iterate over the map.

I consider that to be outside the scope of this PR, though.